AppCenterDataStorage/AppCenterDataStorage/Internal/Client/MSCoreDataClient.m

```diff
@@ -0,0 +1,235 @@
+                                                                                           lastUpdatedDate:cachedDocument.lastUpdatedDate
+                                                                                          pendingOperation:operation
+                                                                                                     error:nil];
+
+        // Update local store and return document.
+        [self updateLocalStore:token
+            currentCachedDocument:cachedDocument
+                newCachedDocument:createdOrUpdatedDocument
+                 deviceTimeToLive:deviceTimeToLive
+                        operation:operation];
+        completionHandler(createdOrUpdatedDocument);
+      }
+    }
+  });
+}
+
+/**
+ * Validate an operation.
+ *
+ *@param operation The operation.
+ *
+ *@return YES if the operation is valid; NO otherwise.
+ */
++ (BOOL)isValidOperation:(NSString *)operation {
+  return operation == nil || [kMSPendingOperationCreate isEqualToString:operation] ||
+         [kMSPendingOperationReplace isEqualToString:operation] || [kMSPendingOperationDelete isEqualToString:operation];
+}
+
+/**
+ * Returns a flag indicating if a remote operation should be attempted.
+ *
+ * @param cachedDocument The current cached document (if any).
+ *
+ * @return YES if a remote operation should be attempted; NO otherwise.
+ */
+- (BOOL)needsRemoteOperation:(MSDocumentWrapper *)cachedDocument {
+  return // Not offline
+      [self.reachability currentReachabilityStatus] != NotReachable
+      // And no pending operation on cached document
+      && cachedDocument.pendingOperation == nil;
+}
+
+- (void)updateLocalStore:(MSTokenResult *)token
+    currentCachedDocument:(MSDocumentWrapper *)currentCachedDocument
+        newCachedDocument:(MSDocumentWrapper *)newCachedDocument
+         deviceTimeToLive:(NSInteger)deviceTimeToLive
+                operation:(NSString *_Nullable)operation {
+
+  // If the device time to live does not allow it, do not touch the local storage.
+  if (deviceTimeToLive == MSDataStoreTimeToLiveNoCache) {
+    return;
+  }
+
+  // If the cached document has a create or replace pending operation, no etags and if the current operation is a
+  // deletion, delete the document from the store.
+  if ([self isPendingOperationNeverSyncedForDocument:cachedDocument currentOperation:operation]) {
+    MSLogInfo([MSDataStore logTag], @"Removing never-synced document from local storage");
+    [self.documentStore deleteWithToken:token documentId:currentCachedDocument.documentId];
+  }
+
+  // Update document storage.
+  else {
+    MSLogInfo([MSDataStore logTag], @"Updating/inserting document into local storage");
+    [self.documentStore upsertWithToken:token documentWrapper:newCachedDocument operation:operation deviceTimeToLive:deviceTimeToLive];
+  }
+}
+
++ (void)isPendingOperationNeverSyncedForDocument:(MSDocumentWrapper *)cachedDocument
```

Comment:
Can this logic be inlined instead of in a helper method? I don't really understand the name so it is making the code harder for me to understand.

Reply:
The check is to see if there's a pending operation that was never synced.  We see if it wasn't synced by checking if there is no eTag that already exists for the document.  This is used when we're trying to delete a document to see if we can discard the pending operation.

Does the name make sense still?  Do you have a better suggestion.